File: src/Subsystems/Arm.cpp
#include <Robot.h>
#include "Arm.h"

Arm::Arm() {
    this->armMotor = new WPI_TalonSRX(ARM_CIM);
    this->intake = new Intake();
    this->intakeClosed = true;
    this->intakeKicked = false;
    this->ToggleIntake();
    this->cimcoder = new Encoder(WINCH_CIMCODER_A, WINCH_CIMCODER_B);
    this->cimcoder->SetDistancePerPulse(WINCH_DIST_PER_PULSE);
    this->cimcoder->Reset();
    this->targetPos = cimcoder->GetDistance();
}

void Arm::Loop() {
    if(!Robot::oi->joy1->GetRawButton(5) && !Robot::oi->joy1->GetRawButton(3) && fabs(this->GetCurrent()) < 30) {
        if (fabs(this->cimcoder->GetDistance() - this->targetPos) < .5) {
            this->armMotor->Set(0);
        } else if (this->targetPos - this->cimcoder->GetDistance() > 0) {
            this->armMotor->Set(1);
        } else {
            this->armMotor->Set(-1);
        }
    }
    if(fabs(this->GetCurrent()) > 30) {
        this->armMotor->Set(0);
    }
}

void Arm::MoveTo(double position) {
    this->targetPos = position;
}

void Arm::ToggleIntake() {
    this->intakeClosed = this->intake->SetState(!intakeClosed);
}

void Arm::ToggleKick() {
    this->intakeKicked = this->intake->SetKicked(!intakeKicked);
}

void Arm::Close() {
    this->intakeClosed = this->intake->SetState(true);
}

void Arm::Open() {
    this->intakeClosed = this->intake->SetState(false);
}

void Arm::KickDown() {
    this->intakeKicked = this->intake->SetKicked(true);
}

void Arm::KickUp() {
    this->intakeKicked = this->intake->SetKicked(false);
}

double Arm::GetCurrent() {
    return this->armMotor->GetOutputCurrent();
}

File: src/main/java/frc/team2410/robot/Subsystems/Arm.java
package frc.team2410.robot.Subsystems;

import frc.team2410.robot.Robot;
import frc.team2410.robot.RobotMap;
import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
import edu.wpi.first.wpilibj.Encoder;

public class Arm {
	
	public WPI_TalonSRX armMotor;
	private Intake intake;
	public Encoder cimcoder;
	private double position;
	private boolean intakeClosed;
	private boolean intakeKicked;
	public double targetPos;
	double offset;

	public Arm() {
		this.armMotor = new WPI_TalonSRX(RobotMap.ARM_CIM);
		this.intake = new Intake();
		this.close();
		this.kickUp();
		this.intakeClosed = true;
		this.intakeKicked = false;
		this.toggleIntake();
		this.cimcoder = new Encoder(RobotMap.WINCH_CIMCODER_A, RobotMap.WINCH_CIMCODER_B);
		this.cimcoder.setDistancePerPulse(RobotMap.WINCH_DIST_PER_PULSE);
		this.setPosition(0);
		this.targetPos = cimcoder.getDistance();
	}
	
	public void loop() {
		//keeps elevator within half an inch of target position
		if (Robot.oi.getStick() == 0) {
			double aspeed;
			aspeed = (this.targetPos - this.getPosition())/10;
			if(Math.abs(this.targetPos-this.getPosition()) < 1) aspeed = 0;
			if(aspeed < -1) aspeed = -1;
			if(aspeed > 1) aspeed = 1;
			this.armMotor.set(aspeed);
		}
	}
	
	public void autoLoop() {
		double aspeed;
		aspeed = (this.targetPos - this.getPosition())/10;
		if(Math.abs(this.targetPos-this.getPosition()) < 1) aspeed = 0;
		if(aspeed < -1) aspeed = -1;
		if(aspeed > 1) aspeed = 1;
		this.armMotor.set(aspeed);
	}

	public void moveTo(double pos) {
		this.targetPos = pos;
	}

	public void toggleIntake() {
		this.intakeClosed = this.intake.setState(!intakeClosed);
	}

	public void toggleKick() {
		this.intakeKicked = this.intake.setKicked(!intakeKicked);
	}

	public void close() {
		this.intakeClosed = this.intake.setState(true);
	}

	public void open() {
		this.intakeClosed = this.intake.setState(false);
	}

	public void kickDown() {
		this.intakeKicked = this.intake.setKicked(true);
	}

	public void kickUp() {
		this.intakeKicked = this.intake.setKicked(false);
	}

	public double getCurrent() {
		return this.armMotor.getOutputCurrent();
	}
	
	public double getPosition() {
		return cimcoder.getDistance() + offset;
	}
	
	public void setPosition(double position) {
		cimcoder.reset();
		offset = position;
		moveTo(position);
	}
}
